internal/server/pool_test.go

```diff
@@ -53,7 +54,455 @@ func TestPoolSpike(t *testing.T) {
+	p.ReleaseConn(conn1)
+
+	be.Equal(t, 2, p.Stats().OpenConnections)
+	be.Equal(t, 2, p.Stats().Idle)
+
+	time.Sleep(100 * time.Millisecond)
+
+	be.Equal(t, 2, p.Stats().OpenConnections)
+	be.Equal(t, 2, p.Stats().Idle)
+
+	p.SetConnMaxIdleTime(100 * time.Millisecond)
+
+	time.Sleep(p.maxIdleTime + time.Second)
+
+	be.Equal(t, 0, p.Stats().OpenConnections)
+	be.Equal(t, 0, p.Stats().Idle)
+	be.Equal(t, 2, p.Stats().MaxIdleTimeClosed)
+}
+
+func TestMaxLifeTime(t *testing.T) {
+	t.Run("basic", func(t *testing.T) {
+		cfg := genBasePoolConfig()
+		cfg.MaxOpen = 3
+		cfg.MaxIdle = 3
+		cfg.MaxLifetime = 1 * time.Second
+		cfg.MaxIdleTime = cfg.MaxLifetime * 10
+
+		rt := be.Relaxed(t)
+
+		p, err := NewPool(cfg)
+		be.NilErr(rt, err)
+		defer func() { be.NilErr(rt, p.Close()) }()
+
+		conn0, err := p.AcquireConn()
+		be.NilErr(rt, err)
+
+		conn1, err := p.AcquireConn()
+		be.NilErr(rt, err)
+
+		be.Equal(t, 2, p.Stats().OpenConnections)
+		be.Equal(t, 0, p.Stats().Idle)
+
+		p.ReleaseConn(conn0)
+		p.ReleaseConn(conn1)
+
+		be.Equal(t, 2, p.Stats().OpenConnections)
+		be.Equal(t, 2, p.Stats().Idle)
+
+		time.Sleep(cfg.MaxLifetime + time.Second)
+
+		be.Equal(t, 0, p.Stats().OpenConnections)
+		be.Equal(t, 0, p.Stats().Idle)
+		be.Equal(t, 2, p.Stats().MaxLifetimeClosed)
+	})
+
+	t.Run("release conn", func(t *testing.T) {
+		cfg := genBasePoolConfig()
+		cfg.MaxOpen = 3
+		cfg.MaxIdle = 3
+		cfg.MaxLifetime = 1 * time.Second
+		cfg.MaxIdleTime = cfg.MaxLifetime * 10
+
+		rt := be.Relaxed(t)
+
+		p, err := NewPool(cfg)
+		be.NilErr(rt, err)
+		defer func() { be.NilErr(rt, p.Close()) }()
+
+		conn0, err := p.AcquireConn()
+		be.NilErr(rt, err)
+
+		conn1, err := p.AcquireConn()
+		be.NilErr(rt, err)
+
+		be.Equal(t, 2, p.Stats().OpenConnections)
+		be.Equal(t, 0, p.Stats().Idle)
+
+		time.Sleep(cfg.MaxLifetime + time.Second)
+
+		p.ReleaseConn(conn0)
+		p.ReleaseConn(conn1)
+
+		be.Equal(t, 0, p.Stats().OpenConnections)
+		be.Equal(t, 0, p.Stats().Idle)
+		be.Equal(t, 2, p.Stats().MaxLifetimeClosed)
+	})
+}
+
+func TestMaxLifetimeUpdate(t *testing.T) {
+	cfg := genBasePoolConfig()
+	cfg.MaxOpen = 3
+	cfg.MaxIdle = 3
+	cfg.MaxLifetime = 1 * time.Second
+	cfg.MaxIdleTime = cfg.MaxLifetime * 10
+
+	rt := be.Relaxed(t)
+
+	p, err := NewPool(cfg)
+	be.NilErr(rt, err)
+	defer func() { be.NilErr(rt, p.Close()) }()
+
+	conn0, err := p.AcquireConn()
+	be.NilErr(rt, err)
+
+	conn1, err := p.AcquireConn()
+	be.NilErr(rt, err)
+
+	be.Equal(t, 2, p.Stats().OpenConnections)
+	be.Equal(t, 0, p.Stats().Idle)
+
+	p.ReleaseConn(conn0)
+	p.ReleaseConn(conn1)
+
+	be.Equal(t, 2, p.Stats().OpenConnections)
+	be.Equal(t, 2, p.Stats().Idle)
+
+	time.Sleep(100 * time.Millisecond)
+
+	be.Equal(t, 2, p.Stats().OpenConnections)
+	be.Equal(t, 2, p.Stats().Idle)
+
+	p.SetConnMaxLifetime(100 * time.Millisecond)
+
+	time.Sleep(p.maxLifetime + time.Second)
+
+	be.Equal(t, 0, p.Stats().OpenConnections)
+	be.Equal(t, 0, p.Stats().Idle)
+	be.Equal(t, 2, p.Stats().MaxLifetimeClosed)
+}
```

Comment:
Same comment as before, but here's not a nit anymore if I'm not mistaken, as we're sleeping for more than the original `MaxLifeTime`, right?

Reply:
Yep right